### PR TITLE
lesson04 -07

### PR DESCRIPTION
**Pull request by luftsmeerflier**
Adding lesson04 mailroom_2.py and lesson07 exercises

### CODE REVIEW

File: students/luftsmeerflier/lesson04/mailroom_2.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	]
+}
+
+def send_thank_you(first_name='', last_name=''):
+	names_list = list(donors.keys())
+
+	_list = input("Type `list` for a list of donors or press the carriage key to skip\n")
+	if _list == "list":
+		for donor in names_list:
+			print(donor)
```
Comment (natasha-aleksandrova) [inline students/luftsmeerflier/lesson04/mailroom_2.py:33]: your indentation is incorrect - should be exactly 4 spaces

File: students/luftsmeerflier/lesson04/mailroom_2.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+		print('\n')
+	else:
+		pass
+
+	name = input('What is the name of the donor?\n').title()
+
+	if name in names_list:
```
Comment (natasha-aleksandrova) [inline students/luftsmeerflier/lesson04/mailroom_2.py:40]: you want to utilize dictionary look up which is more efficient than list look up.

```
donors.get(name)
```

File: students/luftsmeerflier/lesson04/mailroom_2.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		index = names_list.index(name)
+		name = names_list[index]
+		donor = donors[name]
+		date = donor[0][0]
+		amount = donor[0][1]
+		write_letters(name, amount, date)
```
Comment (natasha-aleksandrova) [inline students/luftsmeerflier/lesson04/mailroom_2.py:46]: this code is missing functionality - ability to add / update donations

File: students/luftsmeerflier/lesson04/mailroom_2.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	
+
+def write_letters(name, amount, date):	
+
+	text = "Dear {},\n\n    Thank you for your donation in the amount of {} on {}. Please keep this letter for your records.\n\n"
+
+	letterhead = "{}.txt".format(name).replace(" ", "_")
+	# open a letter template
+	formatted = text.format(name, amount, date)
+	thank_you = open("{}".format(letterhead), "w+")
```
Comment (natasha-aleksandrova) [inline students/luftsmeerflier/lesson04/mailroom_2.py:56]: this should be using `with` statement which closes the file for you

File: students/luftsmeerflier/lesson04/mailroom_2.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+def create_report():
+	#print(donor_list[0].ljust(20) + donor_list[1].ljust(10) + donor_list[2].ljust(20) + donor_list[3].ljust(8))
+	# print(donor_list[0].ljust(20))
+	print('\n')
+	print("Donor Name".ljust(28) + '|'.ljust(2) + "Total Given" + '|'.rjust(2) + "Num Gifts".rjust(10) + "|".ljust(2) + "Average Gift")
+	print('-' * 70)
+
+	names_list = list(donors.keys())
+	names_list.sort()
```
Comment (natasha-aleksandrova) [inline students/luftsmeerflier/lesson04/mailroom_2.py:72]: the sorting functionality isn't working as expected - needs to be sorted by total given.

File: students/luftsmeerflier/lesson04/mailroom_2.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	
+def exit_program():
+	print("Adieu, auf wiedersehen, goodbye!")
+	sys.exit()
+
+arg_dict = {
+    1: send_thank_you,
+    2: create_report,
+	3: send_letter_all,
+	4: exit
```
Comment (natasha-aleksandrova) [inline students/luftsmeerflier/lesson04/mailroom_2.py:115]: your indentation is incorrect here.

File: students/luftsmeerflier/lesson04/mailroom_2.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
+#/usr/bin/env python3
+import sys
+
+donors = {
+	"John Quigley": [
+		("2019-02-04", 5000.00),
+		("2019-02-04", 4250.00),
+		("2019-02-04", 3000.00)
```
Comment (natasha-aleksandrova) [inline students/luftsmeerflier/lesson04/mailroom_2.py:8]: You can simplify this structure and use a dictionary:
`{"Bob": {"2019-02-04": 5000, "2019-02-04": 4250}}`

then you can get all donations by values `donors.values()`and make your code much simpler

**Comment by natasha-aleksandrova** [comment]
I'm going to go ahead and close this for now.

**Comment by luftsmeerflier** [comment]
Natasha,
I made the revisions to lesson04 mailroom_2 according to your comments.
However, I did not find any instructions to add donation entries:
https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/mailroom-part2.html

I look forward to your comments.

-John

**Comment by natasha-aleksandrova** [comment]
@luftsmeerflier each version of the mailroom builds on top of the previous - so if mailroom version 1 had feature to add donation then your version 2 should continue to have that same feature.

let me know if you have any further questions.

**Comment by luftsmeerflier** [comment]
Natasha,
I've added the ability to add a donor to the data structure:
https://github.com/UWPCE-PythonCert-ClassRepos/SP_Online_PY210/compare/master...luftsmeerflier:master

The new code starts on line 30.

**Pull request closed without merging**